lib/widget/toggle_button.dart

```diff
@@ -0,0 +1,303 @@
+      'onChange': (definition) => _controller.onChange =
+          framework.EnsembleAction.fromYaml(definition, initiator: this),
+    };
+  }
+
+  @override
+  Map<String, Function> methods() {
+    return {};
+  }
+
+  bool isValueInItems() {
+    if (_controller.maybeValue != null && _controller.items != null) {
+      for (SelectOneItem item in _controller.items!) {
+        if (_controller.maybeValue == item.value) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  dynamic getValue() {
+    if (isValueInItems()) {
+      return _controller.maybeValue;
+    }
+    return null;
+  }
+
+  void updateItems(dynamic values) {
+    List<SelectOneItem> entries = [];
+    if (values is List) {
+      calculateIconSize(values);
+      for (var element in values) {
+        // must be of value/label pair. Maybe let user overrides later
+        if (element is Map) {
+          if (element['value'] != null) {
+            entries.add(
+              SelectOneItem(
+                value: element['value'],
+                label: element['label']?.toString(),
+                icon: Utils.getIcon(element['icon']),
+                isIcon: isIconExist(values),
+              ),
+            );
+          }
+        }
+        // simply use the value
+        else {
+          entries.add(SelectOneItem(value: element));
+        }
+      }
+    }
+    _controller.items = entries;
+
+    // ensure that the value is still correct
+    if (!isValueInItems()) {
+      _controller.maybeValue = null;
+    }
+  }
+
+  bool isIconExist(List v) {
+    for (var e in v) {
+      if (e['icon'] != null) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void calculateIconSize(List v) {
+    List iconSize = [];
+    _controller.gap = 0;
+    for (var e in v) {
+      if (e is Map) {
+        if (e['icon'] != null) {
+          if (e['icon']['size'] != null) {
+            iconSize.add(e['icon']['size']);
+          }
+        }
+      }
+    }
+    if (iconSize.isNotEmpty) {
+      _controller.gap =
+          iconSize.reduce((curr, next) => curr > next ? curr : next);
+    } else {
+      _controller.gap = 0;
+    }
+  }
+
+  void onSelectionChanged(dynamic value) {
+    setProperty('value', value);
+  }
+}
+
+class EnsembleToggleButtonState extends WidgetState<EnsembleToggleButton> {
+  List<ToggleItem>? _items = [];
+
+  @override
+  void initState() {
+    super.initState();
+    _setValue();
+  }
+
+  void _setValue() {
+    if (widget.controller.items != null) {
+      final value = widget.getValue();
+      final valueIndex = widget.controller.items
+          ?.indexWhere((element) => element.value == value);
+      widget.controller.items!.asMap().forEach((index, value) {
+        _items!.add(
+          ToggleItem(
+            label: value.label,
+            value: value.value,
+            icon: value.icon,
+            isIcon: value.isIcon,
+            isSelected: (valueIndex != null && valueIndex != -1)
```

Comment:
nit: `index` is never null or -1 so you don't need this check

Reply:
I'm checking the value is present in the items (list of type SelectOneItem) and returning its index using the `indexWhere` method. It returns -1 if there are no items found and the controller items are nullable lists, so you'll get an optional value here. It's better to handle it safely